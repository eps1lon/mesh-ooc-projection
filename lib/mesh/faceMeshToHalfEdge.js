"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HalfEdge_1 = require("./half_edge/HalfEdge");
const DoubleLinkedList_1 = require("./half_edge/util/DoubleLinkedList");
// TODO: validate that prev,next,inverse are set
function facesToHalfEdges(mesh) {
    const edges = [];
    const faces = [];
    const vertices = mesh.vertices.map((vertex, id) => ({
        id,
        p: vertex.p,
        attributes: vertex.attributes
    }));
    let edge_id = 0;
    mesh.faces.forEach((mesh_face, face_id) => {
        const face = {
            id: face_id
        };
        const half_edges = [];
        // build half edges
        mesh_face.vertices.forEach((cur_vertex_id, i) => {
            const next_vertex_index = (i + 1) % mesh_face.vertices.length;
            const next_vertex_id = mesh_face.vertices[next_vertex_index];
            const next_vertex = vertices[next_vertex_id];
            const half_edge = {
                id: edge_id,
                target: next_vertex,
                face
            };
            // first generated edge gets referenced in face
            if (face.edge === undefined) {
                face.edge = half_edge;
            }
            // vertex doesnt have an edge yet
            if (vertices[cur_vertex_id].edge === undefined) {
                vertices[cur_vertex_id].edge = half_edge;
            }
            ++edge_id;
            half_edges.push(half_edge);
        });
        faces.push(face);
        edges.push(...DoubleLinkedList_1.doubleLink(half_edges));
    });
    const { non_manifold } = inverseMatching(edges);
    if (non_manifold > 0) {
        // tslint:disable-next-line: no-console
        console.warn('encountered non manifold edges, pathfinding might not work');
    }
    // mark every vertex with boundary edges as incomplete (see #9)
    for (const vertex of vertices) {
        vertex.attributes.boundary = HalfEdge_1.outgoingEdges(vertex).some(edge => edge.inverse === undefined);
    }
    // at this point every vertex is initialized
    return { edges, faces, vertices };
}
exports.default = facesToHalfEdges;
function inverseMatching(half_edges) {
    let borders = 0;
    let matches = 0;
    let non_manifold = 0;
    const sorted = half_edges
        .map(edge => {
        return [HalfEdge_1.origin(edge).id, edge.target.id, edge];
    })
        .map(([source, target, edge]) => {
        // internal index sort
        if (source > target) {
            return [target, source, edge];
        }
        else {
            return [source, target, edge];
        }
    })
        .sort((a, b) => {
        const primary = a[0] - b[0];
        const secondary = a[1] - b[1];
        return primary !== 0 ? primary : secondary;
    });
    for (let i = 0; i < sorted.length - 1;) {
        const edge = sorted[i][2];
        let j = i + 1;
        for (j; j < sorted.length; ++j) {
            if (sorted[i][0] !== sorted[j][0] || sorted[i][1] !== sorted[j][1]) {
                break;
            }
        }
        const pairs = j - i;
        if (pairs === 1) {
            edge.boundary = true;
            edge.manifold = true;
            ++borders;
        }
        else if (pairs === 2) {
            const inverse = sorted[i + 1][2];
            edge.inverse = inverse;
            edge.manifold = true;
            edge.boundary = false;
            inverse.inverse = edge;
            inverse.manifold = false;
            inverse.boundary = false;
            matches = matches + 2;
        }
        else {
            for (let k = i; k < j; ++k) {
                sorted[k][2].manifold = false;
                sorted[k][2].boundary = false;
            }
            non_manifold += pairs;
        }
        i = i + pairs;
    }
    return { matches, borders, non_manifold };
}
