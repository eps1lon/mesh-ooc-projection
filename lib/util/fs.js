"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const util_1 = require("util");
exports.readFile = util_1.promisify(fs.readFile);
exports.writeFile = util_1.promisify(fs.writeFile);
exports.readdir = util_1.promisify(fs.readdir);
