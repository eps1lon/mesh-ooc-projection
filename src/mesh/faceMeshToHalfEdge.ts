import FaceMesh, { VertexId } from './FaceMesh';
import {
  HalfEdgeMesh,
  HalfEdge,
  Face,
  Vertex,
  outgoingEdges,
  origin
} from './half_edge/HalfEdge';
import { doubleLink } from './half_edge/util/DoubleLinkedList';

// TODO: validate that prev,next,inverse are set
export default function facesToHalfEdges(mesh: FaceMesh): HalfEdgeMesh {
  const edges: HalfEdge[] = [];
  const faces: Face[] = [];
  const vertices: Vertex[] = mesh.vertices.map(
    (vertex, id) =>
      ({
        id,
        p: vertex.p,
        attributes: vertex.attributes
      } as Vertex)
  );

  let edge_id = 0;
  mesh.faces.forEach((mesh_face, face_id) => {
    const face: Face = {
      id: face_id
    } as Face;

    const half_edges: HalfEdge[] = [];

    // build half edges
    mesh_face.vertices.forEach((cur_vertex_id, i) => {
      const next_vertex_index = (i + 1) % mesh_face.vertices.length;
      const next_vertex_id = mesh_face.vertices[next_vertex_index];
      const next_vertex = vertices[next_vertex_id];

      const half_edge: HalfEdge = {
        id: edge_id,
        target: next_vertex,
        face
      } as HalfEdge;

      // first generated edge gets referenced in face
      if (face.edge === undefined) {
        face.edge = half_edge;
      }

      // vertex doesnt have an edge yet
      if (vertices[cur_vertex_id].edge === undefined) {
        vertices[cur_vertex_id].edge = half_edge;
      }

      ++edge_id;

      half_edges.push(half_edge);
    });

    faces.push(face as Face);

    edges.push(...doubleLink(half_edges));
  });

  const { non_manifold } = inverseMatching(edges);
  if (non_manifold > 0) {
    // tslint:disable-next-line: no-console
    console.warn('encountered non manifold edges, pathfinding might not work');
  }

  // mark every vertex with boundary edges as incomplete (see #9)
  for (const vertex of vertices) {
    vertex.attributes.boundary = outgoingEdges(vertex).some(
      edge => edge.inverse === undefined
    );
  }
  // at this point every vertex is initialized
  return { edges, faces, vertices };
}

function inverseMatching(half_edges: HalfEdge[]) {
  let borders = 0;
  let matches = 0;
  let non_manifold = 0;

  const sorted = half_edges
    .map(edge => {
      return [origin(edge).id, edge.target.id, edge] as [
        VertexId,
        VertexId,
        HalfEdge
      ];
    })
    .map(([source, target, edge]) => {
      // internal index sort
      if (source > target) {
        return [target, source, edge] as [VertexId, VertexId, HalfEdge];
      } else {
        return [source, target, edge] as [VertexId, VertexId, HalfEdge];
      }
    })
    .sort((a, b) => {
      const primary = a[0] - b[0];
      const secondary = a[1] - b[1];

      return primary !== 0 ? primary : secondary;
    });

  for (let i = 0; i < sorted.length - 1; ) {
    const edge: HalfEdge = sorted[i][2];

    let j = i + 1;
    for (j; j < sorted.length; ++j) {
      if (sorted[i][0] !== sorted[j][0] || sorted[i][1] !== sorted[j][1]) {
        break;
      }
    }

    const pairs = j - i;

    if (pairs === 1) {
      edge.boundary = true;
      edge.manifold = true;

      ++borders;
    } else if (pairs === 2) {
      const inverse: HalfEdge = sorted[i + 1][2];

      edge.inverse = inverse;
      edge.manifold = true;
      edge.boundary = false;

      inverse.inverse = edge;
      inverse.manifold = false;
      inverse.boundary = false;

      matches = matches + 2;
    } else {
      for (let k = i; k < j; ++k) {
        sorted[k][2].manifold = false;
        sorted[k][2].boundary = false;
      }

      non_manifold += pairs;
    }

    i = i + pairs;
  }

  return { matches, borders, non_manifold };
}
