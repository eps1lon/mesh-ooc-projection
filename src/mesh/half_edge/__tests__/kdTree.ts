/// <reference types="@types/jest" />
/// <reference types="@types/node" />
import * as fs from 'fs';
import * as path from 'path';

import buildKdTree, { Node, isInternal } from '../kdTree';
import faceMeshToHalfEdge from '../../faceMeshToHalfEdge';
import big_mesh from './__fixtures__/mesh';

const serializeNode = (node?: Node) => {
  if (node === undefined) {
    return node;
  }

  return {
    ...node,
    left: serializeNode(isInternal(node) ? node.left : undefined),
    right: serializeNode(isInternal(node) ? node.right : undefined),
    triangles: node.triangles.map(t => t.id)
  };
};

it('should match snapshots', () => {
  const tree = buildKdTree(faceMeshToHalfEdge(big_mesh).faces);
  expect(serializeNode(tree)).toMatchSnapshot();
});
