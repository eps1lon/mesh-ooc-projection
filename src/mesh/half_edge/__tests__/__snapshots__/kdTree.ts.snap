// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshots 1`] = `
Object {
  "bbox": Object {
    "max": Array [
      -18.08405,
      -301.5033,
      107.7483,
    ],
    "min": Array [
      -23.75085,
      -312.9892,
      107.4102,
    ],
  },
  "left": Object {
    "bbox": Object {
      "max": Array [
        -21.68603,
        -308.4229,
        107.7246,
      ],
      "min": Array [
        -23.75085,
        -312.4767,
        107.4795,
      ],
    },
    "left": Object {
      "bbox": Object {
        "max": Array [
          -21.68603,
          -309.7769,
          107.7246,
        ],
        "min": Array [
          -23.4891,
          -312.4767,
          107.4795,
        ],
      },
      "left": Object {
        "bbox": Object {
          "max": Array [
            -22.13051,
            -309.7769,
            107.7121,
          ],
          "min": Array [
            -23.37333,
            -312.4767,
            107.4795,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          42,
          56,
          52,
          53,
          55,
          58,
          48,
          68,
        ],
      },
      "right": Object {
        "bbox": Object {
          "max": Array [
            -21.68603,
            -310.4781,
            107.7246,
          ],
          "min": Array [
            -23.4891,
            -312.4767,
            107.6031,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          64,
          60,
          63,
          62,
          65,
          57,
          61,
          67,
          59,
        ],
      },
      "triangles": Array [
        67,
        68,
        63,
        64,
        62,
        65,
        60,
        61,
        56,
        58,
        59,
        57,
        48,
        55,
        53,
        52,
        42,
      ],
    },
    "right": Object {
      "bbox": Object {
        "max": Array [
          -21.71572,
          -308.4229,
          107.7151,
        ],
        "min": Array [
          -23.75085,
          -310.7228,
          107.4795,
        ],
      },
      "left": Object {
        "bbox": Object {
          "max": Array [
            -22.32691,
            -308.4229,
            107.7121,
          ],
          "min": Array [
            -23.64511,
            -310.7228,
            107.4795,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          36,
          43,
          41,
          45,
          31,
          32,
          39,
          19,
          20,
        ],
      },
      "right": Object {
        "bbox": Object {
          "max": Array [
            -21.71572,
            -308.568,
            107.7151,
          ],
          "min": Array [
            -23.75085,
            -310.4771,
            107.5073,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          30,
          25,
          21,
          38,
          35,
          26,
          40,
          23,
          22,
        ],
      },
      "triangles": Array [
        43,
        40,
        41,
        45,
        38,
        35,
        39,
        32,
        31,
        36,
        30,
        25,
        26,
        21,
        23,
        22,
        20,
        19,
      ],
    },
    "triangles": Array [
      35,
      40,
      38,
      25,
      19,
      30,
      67,
      63,
      20,
      41,
      42,
      48,
      57,
      60,
      58,
      55,
      39,
      31,
      68,
      62,
      21,
      32,
      43,
      52,
      64,
      65,
      45,
      36,
      56,
      53,
      22,
      23,
      61,
      59,
      26,
    ],
  },
  "right": Object {
    "bbox": Object {
      "max": Array [
        -18.08405,
        -301.5033,
        107.7483,
      ],
      "min": Array [
        -22.4844,
        -312.9892,
        107.4102,
      ],
    },
    "left": Object {
      "bbox": Object {
        "max": Array [
          -21.33407,
          -306.645,
          107.7332,
        ],
        "min": Array [
          -22.4844,
          -312.9892,
          107.5216,
        ],
      },
      "left": Object {
        "bbox": Object {
          "max": Array [
            -21.33407,
            -308.9722,
            107.6961,
          ],
          "min": Array [
            -22.4844,
            -310.4781,
            107.5216,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          47,
          49,
          51,
          50,
          46,
          44,
          37,
          28,
          33,
        ],
      },
      "right": Object {
        "bbox": Object {
          "max": Array [
            -21.40649,
            -306.645,
            107.7332,
          ],
          "min": Array [
            -22.45631,
            -312.9892,
            107.5305,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          34,
          14,
          29,
          27,
          24,
          69,
          66,
          54,
          70,
        ],
      },
      "triangles": Array [
        70,
        69,
        66,
        54,
        49,
        51,
        50,
        46,
        44,
        47,
        37,
        34,
        33,
        28,
        29,
        27,
        24,
        14,
      ],
    },
    "right": Object {
      "bbox": Object {
        "max": Array [
          -18.08405,
          -301.5033,
          107.7483,
        ],
        "min": Array [
          -22.11425,
          -306.8357,
          107.4102,
        ],
      },
      "left": Object {
        "bbox": Object {
          "max": Array [
            -18.29505,
            -301.5033,
            107.7385,
          ],
          "min": Array [
            -20.06145,
            -306.4738,
            107.4102,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          0,
          7,
          2,
          6,
          3,
          4,
          8,
          16,
          17,
        ],
      },
      "right": Object {
        "bbox": Object {
          "max": Array [
            -18.08405,
            -302.3419,
            107.7483,
          ],
          "min": Array [
            -22.11425,
            -306.8357,
            107.4703,
          ],
        },
        "left": undefined,
        "right": undefined,
        "triangles": Array [
          18,
          1,
          15,
          13,
          5,
          11,
          9,
          10,
          12,
        ],
      },
      "triangles": Array [
        15,
        16,
        18,
        17,
        13,
        12,
        10,
        11,
        9,
        7,
        8,
        5,
        4,
        3,
        6,
        2,
        1,
        0,
      ],
    },
    "triangles": Array [
      33,
      44,
      49,
      46,
      66,
      69,
      70,
      37,
      34,
      28,
      54,
      51,
      14,
      15,
      24,
      27,
      50,
      47,
      29,
      0,
      4,
      5,
      2,
      3,
      1,
      7,
      16,
      17,
      13,
      9,
      10,
      12,
      6,
      8,
      18,
      11,
    ],
  },
  "triangles": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
  ],
}
`;
